packages/astro/src/assets/internal.ts

```diff
@@ -112,19 +133,20 @@ export async function generateImage(
 			serverRoot
 		)
 	);
+
+	const imageService = (await getConfiguredImageService()) as LocalImageService;
 	const resultData = await imageService.transform(
 		fileData,
 		{ ...options, src: originalImagePath },
 		buildOpts.settings.config.image.service.config
 	);
 
-	const finalFileURL = new URL('.' + filepath, clientRoot);
-	const finalFolderURL = new URL('./', finalFileURL);
-
 	await fs.promises.mkdir(finalFolderURL, { recursive: true });
-	await fs.promises.writeFile(finalFileURL, resultData.data);
+	await fs.promises.writeFile(cachedFileURL, resultData.data);
+	await fs.promises.copyFile(cachedFileURL, finalFileURL);
```

Comment:
Are these FS operations safe? Do you know if errors are handled already?

Reply:
Same answer as the other comment, we don't handle FS errors in build in general, we'd just rethrow it really